### ISSUE TITLE
Intensity adjustement should preserve original slope profile for "sim speed" to work correctly

### ISSUE THREAD

**Issue by mattipee** [GoldenCheetah/GoldenCheetah#3807]
With intensity adjustment for slope workouts as it stands, after course rewriting, the sim only sees the rewritten grade. If you tell the trainer "do 5% instead of 10%", then the sim still needs to know the original 10% in order to make faithful progress along the original course.

The PR I intend to raise against this I hope will also be a useful precursor to #3749.

See also comment https://github.com/GoldenCheetah/GoldenCheetah/issues/3749#issuecomment-770067824
See also https://groups.google.com/g/golden-cheetah-users/c/G7yYcXf-UD4/m/6zGhwuSoAAAJ

**Comment by amtriathlon** [comment]
To put this in context: when workout rewrite was introduced slope mode was just distance and slope driven by device speed: not altitude, gps coordinates or simulation, so a similar approach to erg mode was reasonable.

Now we have much more rich slope mode workouts and the course rewrite is not a good match anymore, specially when using simulated speed.

I think we need to find a clean, and ideally simple, solution to make the intensity change consistent in slope mode, when simulated speed is enabled and not.

**Comment by mattipee** [comment]
Without simulated speed:
- fundamentally, progress is made by trainer speed, not power output
- scaling grade down means trainer speed is "cheap" and unfair progress is made
- scaling grade up means trainer speed is "hard" and unfairly poor progress is made

Is there anything that can be done to improve that, or do we leave it alone?

**Comment by mattipee** [comment]
While gradient-scaling is a simple concept, I think there are two important things to note:
- without simulated speed enabled and as intensity varies away from 100%, any progress made by trainer wheel speed alone (see "distanceTick") is inherently "unfair" - but I think that's unavoidable.
- I have a singlespeed bike geared as 56:16, and gradient-scaling doesn't allow me to ride with "lower intensity" at 0% gradient.

Nevertheless, before moving further, I still think with "simulated speed" enabled, progress should be made against the original course - that's the fundamental raised in this issue. I think of this as simply a bug-fix to the currently implementation.

My questions for you are:
- Do you agree with this as a first fix, or what is your view?
- Where are your thoughts regarding the virtual gearing concept as an alternative (or additional) approach?
- Have you conceived of something else that I haven't considered?

**Comment by amtriathlon** [comment]
> While gradient-scaling is a simple concept, I think there are two important things to note:
> 
> * without simulated speed enabled and as intensity varies away from 100%, any progress made by trainer wheel speed alone (see "distanceTick") is inherently "unfair" - but I think that's unavoidable.

When simulated speed is disabled the current approach is better: you climb a lessened/hardened climb at corresponding higher/lower speed and that is what is recorded.

> * I have a singlespeed bike geared as 56:16, and gradient-scaling doesn't allow me to ride with "lower intensity" at 0% gradient.

I think a constant % added/subtracted on higher/lower would solve this issue and it is more intuitive, scaling <= 0 slopes by a factor it is not.

> Nevertheless, before moving further, I still think with "simulated speed" enabled, progress should be made against the original course - that's the fundamental raised in this issue. I think of this as simply a bug-fix to the currently implementation.
> 
> My questions for you are:
> 
> * Do you agree with this as a first fix, or what is your view?

Yes, but likely should apply only when simulated speed is enabled, due to the consideration above.

> * Where are your thoughts regarding the virtual gearing concept as an alternative (or additional) approach?

It seems an interesting theoretical problem, but I would like to see a clear description (I mean one an end user also can understand) and a clean implementation doing exactly that to consider it a practical alternative. Anyway this is unrelated to this issue, AFAIK.

> * Have you conceived of something else that I haven't considered?

To increase decrease slope by addition instead of multiplication, similar to manual mode, due to the reasons above.

**Comment by mattipee** [comment]
Where my thoughts are at the moment is that gradient-scaling (ie Zwift "trainer difficulty") and gradient-offset are achieving two different things, and I don't think it;s necessarily a choice or one or the other.

There are different ways to express it, but gradient-scaling means you can complete a given ride with fewer physical gears, and gradient-offset (or for that matter virtual-gears) could then be considered a secondary adjustment to trainer intensity, if you were to, say, remain in a given physical gear, or have insufficient gears for the simulated grade. There is overlap, but they are addressing different things.

In other words, I'd like to think there is a elegant/practical solution that is a combination of the two.

**Comment by mattipee** [comment]
A proposed direction, trying to maintain:
- flexibility (separate and concurrent use of both gradient adjustments)
- usability (wrt configuration and intensity up/down behaviour)
- consistency (use is the same across sim-speed enabled and disabled)

gradient-scaling:
- implement as "trainer difficulty" in a way very similar to Zwift
- with a configuration slider, perhaps in "Options->Training->Preferences"
- can be changed during a workout via preferences dialog, but not via the up/down buttons
- represents an "overall preference", meaning rider uses "fewer" gears over the course of a ride
- [option to scale -ve gradients differently (eg x2 scale factor) - as per Zwift's implementation]

gradient-offset/virtual-gears:
- implemented by the existing up/down buttons during a workout
- can be changed more easily during a workout than "trainer difficulty"
- applied after gradient-scaling
- addresses the need to reactively increase/decrease resistance
- means rider can use a "different" gear at any given moment


trainer_gradient = (workout_gradient * factor) + offset

**Comment by amtriathlon** [comment]
It seems reasonable to me. initially I thought preset gradient-scaling would not be necessary having dynamic gradient-offset, but it may be useful for users not willing or able to use increase/decrease while training.

**Comment by mattipee** [comment]
Thanks, Ale.

I'll work on it.

**Comment by liversedge** [comment]
I'm toying with putting this into the parking lot, is there a pull request or any other development?
No issues if not, just let me know :)

**Comment by mattipee** [comment]
I didn't manage to develop this well enough to PR cleanly, then the weather picked up and I have stopped riding indoors. Perhaps come winter again, but nothing from me now, or soon.

**Comment by amtriathlon** [comment]
I've added a comment to https://github.com/GoldenCheetah/GoldenCheetah/wiki/UG_Workout-Control_General explaining the current behavior: when you lessen or harden a workout that gets reflected in the recorded file. So "sim speed" works correctly for the lessened or hardened workout.
I understand the rationale for this feature request, but I think the title is not accurate.

**Comment by amtriathlon** [comment]
I am closing this due to inactivity.
